lib/common/socket.c

```diff
@@ -1668,39 +1733,72 @@ int h2o_socket_ebpf_init_key(h2o_ebpf_map_key_t *key, void *_sock)
     return h2o_socket_ebpf_init_key_raw(key, sock_type, (void *)&local, (void *)&remote);
 }
 
-h2o_ebpf_map_value_t h2o_socket_ebpf_lookup(h2o_loop_t *loop, int (*init_key)(struct st_h2o_ebpf_map_key_t *key, void *cbdata),
-                                            void *cbdata)
+static int get_return_map_fd(h2o_loop_t *loop)
 {
-    // try open map if not opened
-    open_tracing_map(loop);
+    static __thread int fd = -1;
+    static __thread uint64_t last_attempt = 0;
+    get_map_fd(loop, H2O_EBPF_RETURN_MAP_PATH, &fd, &last_attempt);
+    return fd;
+}
 
-    // map is not connected, fallback accepting probe
-    if (tracing_map_fd < 0)
-        return (h2o_ebpf_map_value_t){0};
+uint64_t h2o_socket_ebpf_lookup_flags(h2o_loop_t *loop, int (*init_key)(h2o_ebpf_map_key_t *key, void *cbdata), void *cbdata)
+{
+    uint64_t flags = 0;
 
-    // lookup map for our key
+    int tracing_map_fd = get_tracing_map_fd(loop);
     h2o_ebpf_map_key_t key;
-    if (!init_key(&key, cbdata))
-        return (h2o_ebpf_map_value_t){0};
-    return lookup_map(&key);
+    if ((tracing_map_fd >= 0 || H2O__PRIVATE_SOCKET_LOOKUP_FLAGS_ENABLED()) && init_key(&key, cbdata)) {
+        if (tracing_map_fd >= 0)
+            ebpf_map_lookup(tracing_map_fd, &key, &flags);
+
+        int return_map_fd;
+        if (H2O__PRIVATE_SOCKET_LOOKUP_FLAGS_ENABLED() && (return_map_fd = get_return_map_fd(loop)) >= 0) {
+            pid_t tid = gettid();
+
+            /* Make sure a possible old flags is not set, otherwise the subsequent logic will be unreliable. */
+            if (ebpf_map_delete(return_map_fd, &tid) == 0 || errno == ENOENT) {
+                H2O__PRIVATE_SOCKET_LOOKUP_FLAGS(tid, flags, &key);
+
+                if (ebpf_map_lookup(return_map_fd, &tid, &flags) != 0) {
+                    if (errno == ENOENT) {
+                        /* ENOENT could be issued in some reasons even if BPF tries to insert the entry, for example:
+                         *  * the entry in LRU hash was evicted
+                         *  * the insert operation in BPF program failed with ENOMEM */
+                        h2o_error_printf(
+                            "BPF_MAP_LOOKUP failed. "
+                            "BPF handler for h2o:socket_lookup_flags_private might not have set the flags via h2o_return map\n");
```

Comment:
As correctly stated by the comment right above, we expect this branch to be taken at some probability. Based on that, I tend to believe that we should be capping the frequency of this message being emitted. Otherwise, we'd have the chance of this message kicking out all other error messages that would be of more interest.

For a predent, please refer to `track_sendmsg` of lib/http3/common.c

Reply:
Nice catch. 84bec57ea

I'm not sure we'd better implement a shared utility for this kind of logs because we could revert 84bec57ea in the future, the time when we know this ENOENT is unlikely to occur and thus we'd like to know raw logs instead of summarized logs.

Reply:
Thank you for the changes. Reviewing the added code, I actually noticed that there's an issue in the existing code that's being copy-pasted.

That indicates that it's difficult to implement correctly, therefore I think we should not only fix the bug but also make it reusable so that we do not fall into the same pitfall. Hence filed #2675.

I think we'd prefer merging #2675 then using that as the bases for implementing rate-limiting this warning.